nemo/collections/asr/parts/submodules/ctc_beam_decoding.py

```diff
@@ -374,81 +377,106 @@ def default_beam_search(
-        #     )  # type: pyctcdecode.BeamSearchDecoderCTC
-        #
-        # x = x.to('cpu')
-        #
-        # with typecheck.disable_checks():
-        #     beams_batch = []
-        #     for sample_id in range(len(x)):
-        #         logprobs = x[sample_id, : out_len[sample_id], :]
-        #         result = self.pyctcdecode_beam_scorer.decode_beams(
-        #             logprobs,
-        #             beam_width=self.beam_size,
-        #             beam_prune_logp=self.pyctcdecode_cfg.beam_prune_logp,
-        #             token_min_logp=self.pyctcdecode_cfg.token_min_logp,
-        #             prune_history=self.pyctcdecode_cfg.prune_history,
-        #             hotwords=self.pyctcdecode_cfg.hotwords,
-        #             hotword_weight=self.pyctcdecode_cfg.hotword_weight,
-        #             lm_start_state=None,
-        #         )  # Output format: text, last_lm_state, text_frames, logit_score, lm_score
-        #         beams_batch.append(result)
-        #
-        # nbest_hypotheses = []
-        # for beams_idx, beams in enumerate(beams_batch):
-        #     hypotheses = []
-        #     for candidate_idx, candidate in enumerate(beams):
-        #         # Candidate = (text, last_lm_state, text_frames, logit_score, lm_score)
-        #         hypothesis = rnnt_utils.Hypothesis(
-        #             score=0.0, y_sequence=[], dec_state=None, timestep=[], last_token=None
-        #         )
-        #
-        #         # TODO: Requires token ids to be returned rather than text.
-        #         pred_token_ids = None
-        #
-        #         hypothesis.y_sequence = pred_token_ids
-        #         hypothesis.text = candidate[0]  # text
-        #         hypothesis.score = candidate[4]  # score
-        #
-        #         if self.preserve_alignments:
-        #             hypothesis.alignments = x[beams_idx][: out_len[beams_idx]]
-        #
-        #         hypotheses.append(hypothesis)
-        #
-        #     hypotheses = rnnt_utils.NBestHypotheses(hypotheses)
-        #     nbest_hypotheses.append(hypotheses)
-        #
-        # return nbest_hypotheses
+        """
+        PyCTCDecode Beam Search Algorithm. Should support Char and Subword models.
+
+        Args:
+            x: Tensor of shape [B, T, V+1], where B is the batch size, T is the maximum sequence length,
+                and V is the vocabulary size. The tensor contains log-probabilities.
+            out_len: Tensor of shape [B], contains lengths of each sequence in the batch.
+
+        Returns:
+            A list of NBestHypotheses objects, one for each sequence in the batch.
+        """
+        if self.compute_timestamps:
+            raise ValueError(
+                f"Beam Search with strategy `{self.search_type}` does not support time stamp calculation!"
+            )
+
+        try:
+            import pyctcdecode
```

Comment:
Is there a particular reason to import in the function? Seems like Flashlight is imported in the function to avoid circular import.

Reply:
All beam search are imported directly inside of the forward in order to prevent runtime import issue if the library is not installed.